modules/objdetect/src/qrcode.cpp

```diff
@@ -2370,4 +4701,1159 @@ bool QRCodeDetector::detectAndDecodeMulti(
+        possible_version.push_back(2);
+        possible_version.push_back(3);
+    }
+    else{
+        possible_version.push_back(3);
+    }
+    return true;
+}
+
+int QREncoder::versionAuto(const std::string & input_str){
+    vector<int> possible_version;
+    versionEstimate((int)input_str.length(),possible_version);
+    int v = 0;
+    vector<uint8_t> payload_tmp;
+    int version_range[5]={0,1,10,27,41};
+    for(size_t i = 0 ; i < possible_version.size() ; i++ ){
+        int version_range_index = possible_version[i];
+        if(version_range_index == 1 ){
+            v = 1;
+        }
+        else if (version_range_index == 2 ){
+            v = 10;
+        }
+        else{
+            v = 27;
+        }
+        encodeAuto(input_str,payload_tmp);
+        v = findVersionCapacity((int)payload_tmp.size(),ecc_level,
+                                            version_range[version_range_index],version_range[version_range_index+1]);
+        if(v!= -1)
+            break;
+    }
+    return v;
+}
+
+QREncoder::QREncoder(const std::string& input ,int mode,int v  = 0 ,int ecc = 0 ,int mask = -1 ,int eci_mode = -1,int structure_num = 2){
+    fnc1_first = false;
+    fnc1_second = false;
+    fnc1_second_AI = 0;
+    mask_type = mask;
+    ecc_level = ecc;
+
+    mode_type = mode;
+    /**use structure mode to split the QRcode into several segmemts*/
+    int struct_num = ( mode == QR_MODE_STRUCTURE ? structure_num:1);
+    /**calculate the parity*/
+    if(struct_num > 1){
+        parity = 0;
+        for(size_t i = 0 ; i < input.length() ; i++ ){
+            parity ^= input[i];
+        }
+        /**can't be over 16*/
+        if(struct_num>16)
+            struct_num = 16;
+        total_num =(uint8_t)struct_num - 1 ;
+
+    }
+
+    int segment_len = (int)ceil((int)input.length()/struct_num);
+    for(int i = 0; i < struct_num ; i ++){
+        sequence_num = (uint8_t)i;
+        /**get the sub string for structure mode*/
+        int segment_begin = i * segment_len;
+        int segemnt_end   = min( (i+1)*segment_len ,(int)input.length()) -1 ;
+        input_info = input.substr(segment_begin,segemnt_end-segment_begin + 1 );
+        /**determine the version automatically*/
+        version_level = (v > 0 ? v : versionAuto(input_info));
+
+        payload.clear();
+        payload.reserve(max_payload_len);//memset(payload,0,sizeof(uint8_t)*max_payload_len);
+        format = Mat(Size(1, 15), CV_8UC1, Scalar(255));
+
+        version_size = (21 + (version_level - 1) * 4);
+        version_info =&version_info_database[version_level];
+        cur_ecc_params = &version_info->ecc[ecc_level];
+
+        /**always use UTF-8 */
+        eci = ((mode_type == QR_MODE_ECI)?eci_mode : -1 );
+
+        original = Mat(Size(version_size, version_size), CV_8UC1, Scalar(255));
+        masked_data = original.clone();
+
+        my_qrcodes.push_back(QRcodeGenerate());
```

Comment:
Because the structure append mode will split the input string into several segments and each segment will generate a qrcode. So the output maybe more than a single Mat .That's why I use a vector to store that.